ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp

```diff
@@ -110,6 +110,340 @@ void max_pool(const T* arg,
+        if (kernel_position[dim] + kernel_offset[dim] < 0 ||
+            kernel_position[dim] + kernel_offset[dim] >= data_shape[dim + 2]) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+template <typename T>
+Coord<T> next_kernel_position_2D(Coord<T> kernel_position,
+                                 const Shape& kernel,
+                                 const Strides& kernel_strides,
+                                 const Strides& kernel_dilations,
+                                 const Shape& data_shape,
+                                 const Shape& pads_begin,
+                                 const Shape& pads_end) {
+    // move the kernel horizontally one stride to the right
+    kernel_position[1] += kernel_strides[1];
+
+    // if the top-right corner of the kernel is outside of the padding area,
+    // move it back to the left and one stride down
+    if (kernel_position[1] + (kernel[1] - 1) * kernel_dilations[1] >= data_shape[3] + pads_end[1]) {
+        kernel_position[1] = 0 - pads_begin[1];
+        kernel_position[0] += kernel_strides[0];
+    }
+
+    return kernel_position;
+}
+
+template <typename T>
+Coord<T> next_kernel_position_3D(Coord<T> kernel_position,
+                                 const Shape& kernel,
+                                 const Strides& kernel_strides,
+                                 const Strides& kernel_dilations,
+                                 const Shape& data_shape,
+                                 const Shape& pads_begin,
+                                 const Shape& pads_end) {
+    kernel_position[2] += kernel_strides[2];
+
+    if (kernel_position[2] + (kernel[2] - 1) * kernel_dilations[2] >= data_shape[4] + pads_end[2]) {
+        kernel_position[2] = 0 - pads_begin[2];
+        kernel_position[1] += kernel_strides[1];
+        if (kernel_position[1] + (kernel[1] - 1) * kernel_dilations[1] >= data_shape[3] + pads_end[1]) {
+            kernel_position[1] = 0 - pads_begin[1];
+            kernel_position[0] += kernel_strides[0];
+        }
+    }
+
+    return kernel_position;
+}
+
+namespace kernel {
+template <typename Values_t, typename Indices_t>
+void max_pool_1d(const Values_t* data,
+                 Values_t* values,
+                 Indices_t* indices,
+                 const size_t data_elems,
+                 const size_t out_elems,
+                 const size_t kernel_size,
+                 const size_t kernel_stride,
+                 const size_t kernel_dilation,
+                 const size_t pads_begin,
+                 const size_t pads_end,
+                 const size_t indices_offset) {
+    int kernel_position = 0 - pads_begin;
+    // select max elem and its index for each "placeholder" in the out buffer (pointed to by out_idx)
+    for (size_t out_idx = 0; out_idx < out_elems; ++out_idx) {
+        Values_t max_elem = std::numeric_limits<Values_t>::lowest();
+        Indices_t max_elem_idx = Indices_t{0};
```

Comment:
Minor optimization. You can initialize the `max_elem_idx` with the index of the element in the top left corner of the kernel and initialize `max_elem` with the value of that element. After that you can iterate `kernel_elem` starting from 1 in line 223 instead of 0.

Reply:
The same comment is applicable for other `max_pool_Xd` functions.

Reply:
I would also recommend to get rid of `max_elem` variable at all. You can use just `max_elem_idx` variable like this: `data[max_elem_idx]` instead of updating two variables each time you find a new max element. CPU cache will keep the value of `max_elem_idx` in some register so accessing it will be very fast.

Reply:
Removing the `max_elem` variable complicates the code a bit because I have to find the first element that doesn't belong to the padding area. Only then I can use its value to initialize `data[max_elem_idx]` for the first time. I can initialize the `max_elem` this way: `Values_t max_elem = kernel_position < 0 ? std::numeric_limits<Values_t>::lowest() : data[kernel_position];` and we'll save one iteration of the inner loop but this will add an if-statement so I'm not sure if it's worth it. 
I'd rather leave it as is to keep the code more readable.